### PR TITLE
feat(python): support moe 

### PR DESCRIPTION
**Pull request by liuhatry**

### CODE REVIEW

File: examples/mnist/main.py [hunk 1]
```diff
@@ -13,14 +13,20 @@
 
 
 class Net(nn.Module):
-    def __init__(self):
+    def __init__(self, num_local_experts):
```
Comment (NOBLES5E) [inline examples/mnist/main.py:16]: move moe example into separate example, for example `examples/mnist-moe`

File: bagua/torch_api/moe/sharded_moe.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
+# The file has been adapted from DeepSpeed:
+#   https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/moe/sharded_moe.py
+# Git commit hash: bff6126f0ddbd1a03da66867571ac87b11c21ac1
+# We retain the following license from the original files:
+
+# Copyright 2020 The Microsoft DeepSpeed Team
```
Comment (NOBLES5E) [inline bagua/torch_api/moe/sharded_moe.py:6]: Also add our license line

File: bagua/torch_api/moe/layer.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
+# The file has been adapted from DeepSpeed:
+#   https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/moe/layer.py
+# Git commit hash: bff6126f0ddbd1a03da66867571ac87b11c21ac1
+# We retain the following license from the original files:
+
+# Copyright 2020 The Microsoft DeepSpeed Team
```
Comment (NOBLES5E) [inline bagua/torch_api/moe/layer.py:6]: add our license

File: bagua/torch_api/moe/experts.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
+# The file has been adapted from DeepSpeed:
+#   https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/moe/experts.py
+# Git commit hash: bff6126f0ddbd1a03da66867571ac87b11c21ac1
+# We retain the following license from the original files:
+
+# Copyright 2020 The Microsoft DeepSpeed Team
```
Comment (NOBLES5E) [inline bagua/torch_api/moe/experts.py:6]: add our license

File: bagua/torch_api/moe/__init__.py [hunk 5]
```diff
@@ -0,0 +1 @@
+from .layer import MoE  # noqa: F401
```
Comment (NOBLES5E) [inline bagua/torch_api/moe/__init__.py:1]: move the whole moe directory to `bagua/torch_api/model_parallel/moe`

**Review by NOBLES5E — CHANGES_REQUESTED** [review]
see comments

File: bagua/torch_api/moe/__init__.py (hunk 5 above)
Comment (liuhatry) [reply bagua/torch_api/moe/__init__.py]: done

File: bagua/torch_api/moe/experts.py (hunk 4 above)
Comment (liuhatry) [reply bagua/torch_api/moe/experts.py]: done

File: bagua/torch_api/moe/layer.py (hunk 3 above)
Comment (liuhatry) [reply bagua/torch_api/moe/layer.py]: done

File: bagua/torch_api/moe/sharded_moe.py (hunk 2 above)
Comment (liuhatry) [reply bagua/torch_api/moe/sharded_moe.py]: done

File: examples/mnist/main.py (hunk 1 above)
Comment (liuhatry) [reply examples/mnist/main.py]: done

**Pull request merged**